python/pyspark/sql/classic/dataframe.py

```diff
@@ -0,0 +1,1952 @@
+                )
+        return self._schema
+
+    def printSchema(self, level: Optional[int] = None) -> None:
+        if level:
+            print(self._jdf.schema().treeString(level))
+        else:
+            print(self._jdf.schema().treeString())
+
+    def explain(
+        self, extended: Optional[Union[bool, str]] = None, mode: Optional[str] = None
+    ) -> None:
+        if extended is not None and mode is not None:
+            raise PySparkValueError(
+                error_class="CANNOT_SET_TOGETHER",
+                message_parameters={"arg_list": "extended and mode"},
+            )
+
+        # For the no argument case: df.explain()
+        is_no_argument = extended is None and mode is None
+
+        # For the cases below:
+        #   explain(True)
+        #   explain(extended=False)
+        is_extended_case = isinstance(extended, bool) and mode is None
+
+        # For the case when extended is mode:
+        #   df.explain("formatted")
+        is_extended_as_mode = isinstance(extended, str) and mode is None
+
+        # For the mode specified:
+        #   df.explain(mode="formatted")
+        is_mode_case = extended is None and isinstance(mode, str)
+
+        if not (is_no_argument or is_extended_case or is_extended_as_mode or is_mode_case):
+            if (extended is not None) and (not isinstance(extended, (bool, str))):
+                raise PySparkTypeError(
+                    error_class="NOT_BOOL_OR_STR",
+                    message_parameters={
+                        "arg_name": "extended",
+                        "arg_type": type(extended).__name__,
+                    },
+                )
+            if (mode is not None) and (not isinstance(mode, str)):
+                raise PySparkTypeError(
+                    error_class="NOT_STR",
+                    message_parameters={"arg_name": "mode", "arg_type": type(mode).__name__},
+                )
+
+        # Sets an explain mode depending on a given argument
+        if is_no_argument:
+            explain_mode = "simple"
+        elif is_extended_case:
+            explain_mode = "extended" if extended else "simple"
+        elif is_mode_case:
+            explain_mode = cast(str, mode)
+        elif is_extended_as_mode:
+            explain_mode = cast(str, extended)
+        assert self._sc._jvm is not None
+        print(self._sc._jvm.PythonSQLUtils.explainString(self._jdf.queryExecution(), explain_mode))
+
+    def exceptAll(self, other: "ParentDataFrame") -> "ParentDataFrame":
+        return DataFrame(self._jdf.exceptAll(other._jdf), self.sparkSession)
+
+    def isLocal(self) -> bool:
+        return self._jdf.isLocal()
+
+    @property
+    def isStreaming(self) -> bool:
+        return self._jdf.isStreaming()
+
+    def isEmpty(self) -> bool:
+        return self._jdf.isEmpty()
+
+    def show(self, n: int = 20, truncate: Union[bool, int] = True, vertical: bool = False) -> None:
```

Comment:
will the docstring be inherited from the superclass?
or still need to copy them one by one? (like `show.__doc__ = PySparkDataFrame.show.__doc__` in connect)

Reply:
It will be inherited:

```
./bin/pyspark --remote local
from pyspark.sql import DataFrame
help(spark.range(1))
```

```
Help on DataFrame in module pyspark.sql.connect.dataframe object:

class DataFrame(pyspark.sql.dataframe.DataFrame)
 |  DataFrame(plan: pyspark.sql.connect.plan.LogicalPlan, session: 'SparkSession') -> 'DataFrame'
 |
 |  Method resolution order:
 |      DataFrame
 |      pyspark.sql.dataframe.DataFrame
 |      builtins.object
 ```

```
./bin/pyspark --master local
from pyspark.sql import DataFrame
help(spark.range(1))
```

```
Help on DataFrame in module pyspark.sql.classic.dataframe object:

class DataFrame(pyspark.sql.dataframe.DataFrame, pyspark.sql.pandas.map_ops.PandasMapOpsMixin, pyspark.sql.pandas.conversion.PandasConversionMixin)
 |  DataFrame(jdf: 'JavaObject', sql_ctx: Union[ForwardRef('SQLContext'), ForwardRef('SparkSession')]) -> 'DataFrame'
 |
 |  Method resolution order:
 |      DataFrame
 |      pyspark.sql.dataframe.DataFrame
 |      pyspark.sql.pandas.map_ops.PandasMapOpsMixin
 |      pyspark.sql.pandas.conversion.PandasConversionMixin
 |      builtins.object
 |
```

but I think I need to double check if doctests run fine.